### PR TITLE
Load data range reorg

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses #586 

Merges `load_data_range` branch into the `inst_reordering` branch. The merge varies from automated to manual, both with support from software (visual indications) as well as visual inspection of code (less reliable).

I did catch a couple deprecation warnings in this branch related to providing keyword `num_samples` for number of simulated samples rather than use `inst_id` as a workaround. Only known difference with `load_data_range` branch.

## Type of change

Please delete options that are not relevant.

- Technically it is a null change, just brings together changes in similar areas done at similar times.

# How Has This Been Tested?

- macOS 10.15.4
- python 3.8.3

# Checklist:

- [ ] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

File: pysat/_instrument.py [hunk 1]
```diff
@@ -1483,106 +1527,272 @@ def bounds(self):
...
-            if end is None:
-                end = self.files.stop_date
-            self._iter_start = [self._filter_datetime_input(start)]
-            self._iter_stop = [self._filter_datetime_input(end)]
-            self._iter_list = utils.time.create_date_range(self._iter_start,
-                                                           self._iter_stop,
-                                                           freq=step)
-            self._iter_type = 'date'
-        else:
-            raise ValueError(''.join(('Provided an invalid combination of',
```
Comment (rstoneback) [inline pysat/_instrument.py:1581]: Note this check is effectively now starting at line 1687.

**Comment by rstoneback** [comment]
I believe this is ready for review @aburrell, pending final review of pull #582.

File: pysat/_instrument.py [hunk 2]
```diff
@@ -3125,3 +3590,42 @@ def _pass_func(*args, **kwargs):
...
+
+def _check_load_arguments_none(*args, raise_error=False):
+    """Ensure all arguments are None.
+
+    Used to support .load method checks that arguments that should be
+    None are None, while also keeping the .load method readable.
+
+    Parameters
+    ----------
+    *args : mixed
```
Comment (aburrell) [inline pysat/_instrument.py:3603]:
```suggestion
    *args : iterable object
```

File: docs/tutorial_basics.rst [hunk 3]
```diff
@@ -194,6 +194,64 @@ type of data structure is flagged at the instrument level with the attribute
...
+.. code:: python
+
+   # load by datetimes
+   dmsp.load(date=datetime.datetime(2001, 1, 1),
+             end_date=datetime.datetime(2001, 1, 3))
+
+   # the following two load commands are equivalent
+   dmsp.load(date=datetime.datetime(2001, 1, 1),
+             end_date=datetime.datetime(2001, 1, 2))
+   dmsp.load(date=datetime.datetime(2001, 1, 1))
```
Comment (aburrell) [inline docs/tutorial_basics.rst:223]: Earlier in this tutorial, datetime was imported using `import datetime as dt`.

**Comment by rstoneback** [comment]
Thanks @aburrell. I incorporated your review comments directly over in #582. I'm pulling them in here via #591 as develop-3 changed in the interim. To make things even more complicated, some of your docstring changes (from #586) are now directly in #582. It's a story.

File: pysat/_orbits.py [hunk 4]
```diff
@@ -552,14 +571,14 @@ def load(self, orbit=None):
...
+                        # check if we are close to beginning or end of day
                         date = self.sat.date
                         delta_start = self.sat.index[-1] - date
                         delta_end = (date + pds.DateOffset(days=1)
                                      - self.sat.index[0])
 
                         if delta_start <= self.orbit_period * 1.05:
-                            # near begining
+                            # near beginning
                             # load previous file, then go forward one orbit
```
Comment (aburrell) [inline pysat/_orbits.py:582]:
```suggestion
                            # We are near the beginning. Load the
                            # previous file, then go forward one orbit
```

File: pysat/tests/test_files.py [hunk 5]
```diff
@@ -33,7 +34,7 @@ def create_dir(inst=None, temporary_file_list=False):
 
 
 def remove_files(inst=None):
-    # remove any files
+    # remove any files like pysat_testing_junk_
     temp_dir = inst.files.data_path
     for the_file in os.listdir(temp_dir):
         if (the_file[0:13] == 'pysat_testing') & \
```
Comment (aburrell) [inline pysat/tests/test_files.py:40]: This will break our flake8 since the operator is before the line break.

File: pysat/_instrument.py [hunk 6]
```diff
@@ -1070,31 +1099,35 @@ def _filter_datetime_input(self, date):
...
+        fid : int or NoneType
+            filename index value (default=None)
+        inc : pds.DateOffset or int
+            Increment of files or dates to load, starting from the
+            root date or fid (default=None)
 
         Returns
-        -------
-        data : (pds.DataFrame or xr.Dataset)
+        --------
```
Comment (aburrell) [inline pysat/_instrument.py:1117]: This looks like it was adjusted to be too long

**Review by aburrell — APPROVED** [review]
I made some comments and suggestions, but the important things are there.

**Comment by rstoneback** [comment]
I looked at the differences between this branch and @aburrell originating reorg using local software. I only see changes related to the expanded data loading. There are more differences shown in the browser given the weird way I merged things in by accident. Good news is everything appears that it will work out once this is in.

**Pull request merged**